lib/types/multipart.js

```diff
@@ -269,34 +271,43 @@ function Multipart (boy, cfg) {
 
 Multipart.prototype.write = function (chunk, cb) {
   let r
-  if ((r = this.parser.write(chunk)) && !this._pause) { cb() } else {
+  if ((r = this.parser.write(chunk)) && !this._pause) {
+    if (!this._boundaryFound && (this._nparts !== 0 || chunk.includes(this._boundary))) {
```

Comment:
Yeah indexOf is usually faster. But the issue is, that imagine having a 1mb chunk. 1024x1024 bytes. If the substring you are searching for is not in the chunk, it will process always to the end, even though you know it has to start with it. 

It is better to modify bufferStartWith to skip whitespace, and then run the startsWith as proposed.

Untested and typed on my mobile while on the way home:
```js
function bufferStartsWith(buf, val) {
  const len = val.length
  let i = 0
  
  // forward whitespace, currently only space, maybe replace with lookup array
  while(buf[i] === 32) { /* space */
    ++i
  }
  if ((buf.length - i)< len) { return false }
  while (i < len) {
    if (buf[i] !== val[i]) { return false }
    ++i
  }
  return true
}
```

Reply:
Should be 

  while(buf[i] === 32 || buf[i] === 9) { /* space and horizontal tab */
    ++i
  }

Reply:
I think we shouldn't skip whitespace per the [RFC](https://www.rfc-editor.org/rfc/rfc2046#section-5.1) section 5.1 and 5.1.1

> The body must then contain
   one or more body parts, each preceded by a boundary delimiter line,
   and the last one followed by a closing boundary delimiter line.

> The boundary delimiter MUST occur at the beginning of a line, i.e.,
   following a CRLF, and the initial CRLF is considered to be attached
   to the boundary delimiter line rather than part of the preceding
   part.  The boundary may be followed by zero or more characters of
   linear whitespace.

I will use your original function

Reply:
cc @KhafraDev if you'd like to review, the benchmarks are still about the same

Reply:
@gurgunday 

How about this:

```js
const DASH = 0x2d /* - */

Multipart.prototype._checkBoundary = function _checkBoundary(chunk) {
  const boundaryLen = this._boundary.length

  if (chunk.length < (boundaryLen + 4)) { return false }

  if (
    chunk[0] !== DASH /* - */ ||
    chunk[1] !== DASH /* - */ ||
    chunk[boundaryLen] !== DASH /* - */ ||
    chunk[boundaryLen + 1] !== DASH /* - */
  ) {
    return false
  }

  const val = Buffer.from(this._boundary)
  for (let i = 2; i < boundaryLen; ++i) {
    if (chunk[i] !== val[i]) { return false }
  }

  return true
}

```

You could use it like this:     

```js
if (this._boundaryNotFound && (this._nparts !== 0 || this._checkBoundary(chunk))) {
```

Reply:
Even better IMO